File: agent/internal/tool/run_command.go
package tool

import (
	"context"
	"fmt"
	"time"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func runCommandTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "run_command",
			Description: "Execute a shell command in the project directory. Use this for builds, tests, linters, or any shell operation.",
			Category:    "system",
			Parameters: map[string]api.ParamDef{
				"command": {Type: "string", Description: "Shell command to execute", Required: true},
				"timeout": {Type: "number", Description: "Timeout in seconds", Required: false, Default: 30},
				"workdir": {Type: "string", Description: "Working directory for the command", Required: false},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			command, _ := params["command"].(string)
			if command == "" {
				return nil, fmt.Errorf("command is required")
			}
			timeoutSec, _ := params["timeout"].(float64)
			if timeoutSec == 0 {
				timeoutSec = 30
			}
			workdir, _ := params["workdir"].(string)

			stdout, stderr, exitCode, _ := runShell(ctx, command, workdir, time.Duration(timeoutSec)*time.Second)

			return map[string]any{
				"stdout":   stdout,
				"stderr":   stderr,
				"exitCode": exitCode,
			}, nil
		},
	}
}

File: src/services/tools/system/runCommand.ts
import { z } from 'zod';
import type { ToolDef } from '../types';

export const runCommandTool: ToolDef = {
  name: 'run_command',
  description: 'Execute a shell command and return its output. Use for running tests, builds, linters, or any CLI tool.',
  category: 'system',
  inputSchema: z.object({
    command: z.string().describe('The shell command to execute (e.g. "npm test", "go build ./...", "python script.py").'),
    workdir: z.string().optional().describe('Working directory for the command. Defaults to the project root.'),
    timeout: z.number().optional().default(30000).describe('Timeout in milliseconds.'),
  }),
  execute: async ({ command, workdir, timeout }) => {
    const backendUrl = import.meta.env.VITE_API_URL || 'http://localhost:3001';
    try {
      const res = await fetch(`${backendUrl}/tools`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ tool: 'run_command', params: { command, workdir, timeout } }),
      });
      if (!res.ok) throw new Error(`Backend error (${res.status})`);
      return res.json();
    } catch (error: any) {
      if (error.message?.includes('fetch')) {
        throw new Error('Command execution requires the Express backend (http://localhost:3001) or the Tauri desktop app');
      }
      throw error;
    }
  },
};
